framework/console/ErrorHandler.php

```diff
@@ -39,7 +39,7 @@ protected function renderException($exception)
             } elseif (count($alternatives) > 1) {
                 $message .= "\n\nDid you mean one of these?\n    - " . implode("\n    - ", $alternatives);
             }
-        } elseif ($exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)) {
+        } elseif ($exception instanceof Exception || !YII_DEBUG) {
```

Comment:
That makes it not display user exceptions in debug mode. Changes behavior.

Reply:
`\yii\console\Exception` is a subclass of `\yii\base\UserException`. If an object is an instance of `\yii\console\Exception`, it must be an instance of `\yii\base\UserException`.  So `$exception instanceof UserException` is redundant.

Reply:
No, it's not.

It will be have differently when `$exception` is `\Exception` and `YII_DEBUG` is `true`. Previously it returned `false` in this case, with your code it's always `true`.

Reply:
The FQCN of  `Exception` in statement `$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)` is `\yii\console\Exception` , not `\Exception`.

When `$exception` is `\Exception` and `YII_DEBUG` is `true`. Previously it returned false in this case, with my code it's also false.

Reply:
Well, I've made a typo then. `\yii\console\Exception` it is where behavior changed.

Reply:
The idea is that user exception should be displayed same way regardless of YII_DEBUG value.

Reply:
Re-checked it. You're right.

`$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)`

|                              | YII_DEBUG=true | YII_DEBUG=false  
|---------------------|-------------------|--------------------
| console Exception | ❌                      | ✅
| UserException       | ✅                      | ✅

`$exception instanceof Exception || !YII_DEBUG`

|                                | YII_DEBUG=true | YII_DEBUG=false  
|----------------------|-------------------|--------------------
| console  Exception | ❌                      | ✅
| UserException        | ✅                      | ✅

Reply:
But `$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)` is no equivalent of `$exception instanceof Exception || !YII_DEBUG`. 

In this case 
`$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)`
could be simplified to:
`$exception instanceof Exception && (true || !YII_DEBUG)`
which could be simplified to:
`$exception instanceof Exception`

Which may suggests that this old condition may be a bug and does not work as intended. 

But nonetheless, this PR changes meaning of this condition.

Reply:
> But nonetheless, this PR changes meaning of this condition.

I've thought so but nope, it does not. See tables above.

Reply:
Oh, I've failed filling the table :) Thanks @rob006. That's what it should be:

`$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)`

|                              | YII_DEBUG=true | YII_DEBUG=false  
|---------------------|-------------------|--------------------
| console Exception | ❌                      | ✅
| UserException       | ✅                      | ✅

`$exception instanceof Exception || !YII_DEBUG`

|                                | YII_DEBUG=true | YII_DEBUG=false  
|----------------------|-------------------|--------------------
| console  Exception | ✅                      | ✅
| UserException        | ❌                      | ✅

Reply:
Fixed the condition: https://github.com/yiisoft/yii2/commit/4a5b325160b04bea5583c390a98054098610b37d

Reply:
@samdark For me both these tables are incorrect :P. But what you really missed is case for `\Exception` and `YII_DEBUG=false`.

Reply:
That's not `\Exception`, that's console exception.

Reply:
Yes. Both are not correct. Should be alright after the fix I've made.

Reply:
@samdark  The table is wrong. It's true when `$exception` is an instance of `\yii\console\Exception` and `YII_DEBUG=true`. `\yii\console\Exception` is a subclass of `\yii\base\UserException`

@rob006 I agree that this redundant condition may be a bug and does not work as intended.

Reply:
@undefinedor - I thought so previously but it's not, just double checked the Samdark's change. Both `$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)` and `$exception instanceof Exception || !YII_DEBUG)` behave the same except one case - when `$exception` is not `UserException` and is not `Exception` and `YII_DEBUG` is `false`:

`$exception instanceof Exception && ($exception instanceof UserException || !YII_DEBUG)` => `F && (F || T)` => `F && T` => `F`
`$exception instanceof Exception || !YII_DEBUG` => `F || T` => `T`